testlapack/dlaqp2.go

```diff
@@ -0,0 +1,132 @@
+// Copyright ©2017 The gonum Authors. All rights reserved.
+// Use of this source code is governed by a BSD-style
+// license that can be found in the LICENSE file.
+
+package testlapack
+
+import (
+	"fmt"
+	"math"
+	"math/rand"
+	"sort"
+	"testing"
+
+	"github.com/gonum/blas"
+	"github.com/gonum/blas/blas64"
+)
+
+type Dlaqp2er interface {
+	Dlaqp2(m, n, offset int, a []float64, lda int, jpvt []int, tau, vn1, vn2, work []float64)
+}
+
+func Dlaqp2Test(t *testing.T, impl Dlaqp2er) {
+	rnd := rand.New(rand.NewSource(1))
+	for ti, test := range []struct {
+		m, n, offset int
+	}{
+		{m: 4, n: 3, offset: 0},
+		// {m: 4, n: 3, offset: 2},
+		// {m: 4, n: 3, offset: 4},
+		{m: 10, n: 10, offset: 0},
+		// {m: 10, n: 10, offset: 5},
+		// {m: 10, n: 10, offset: 10},
+	} {
+		m := test.m
+		n := test.n
+		jpiv := make([]int, n)
+
+		const (
+			all = iota
+			some
+			none
+		)
+		for _, free := range []int{all, some, none} {
+			for _, extra := range []int{0, 11} {
+				a := zeros(m, n, n+extra)
+				c := 1
+				for i := 0; i < m; i++ {
+					for j := 0; j < n; j++ {
+						a.Data[i*a.Stride+j] = float64(c)
+						c++
+					}
+				}
+				aCopy := cloneGeneral(a)
+				for j := range jpiv {
+					switch free {
+					case all:
+						jpiv[j] = -1
+					case some:
+						jpiv[j] = rnd.Intn(2) - 1
+					case none:
+						jpiv[j] = j
+					default:
+						panic("bad freedom")
+					}
+				}
+
+				tau := make([]float64, n)
+				vn1 := make([]float64, n)
+				vn2 := make([]float64, n)
+				work := make([]float64, m)
+
+				impl.Dlaqp2(m, n, test.offset, a.Data, a.Stride, jpiv, tau, vn1, vn2, work)
+
+				prefix := fmt.Sprintf("Case %v (offset=%t,m=%v,n=%v,extra=%v)", ti, test.offset, m, n, extra)
+				if !generalOutsideAllNaN(a) {
+					t.Errorf("%v: out-of-range write to A", prefix)
+				}
+
+				if test.offset == m {
+					continue
+				}
+
+				mo := m - test.offset
+				q := constructQ("QR", mo, n, a.Data[test.offset*a.Stride:], a.Stride, tau)
+				// Check that q is orthonormal
+				for i := 0; i < mo; i++ {
+					nrm := blas64.Nrm2(mo, blas64.Vector{Inc: 1, Data: q.Data[i*mo:]})
+					if math.Abs(nrm-1) > 1e-13 {
+						t.Errorf("Case %v, q not normal", ti)
+					}
+					for j := 0; j < i; j++ {
+						dot := blas64.Dot(mo, blas64.Vector{Inc: 1, Data: q.Data[i*mo:]}, blas64.Vector{Inc: 1, Data: q.Data[j*mo:]})
+						if math.Abs(dot) > 1e-14 {
+							t.Errorf("Case %v, q not orthogonal", ti)
+						}
+					}
+				}
+
+				// Dlaqp2 does not set pivot values, so
+				// we must use column index pivots.
+				if free != none {
+					continue
+				}
+
+				// Check that A * P = Q * R
+				r := blas64.General{
+					Rows:   mo,
+					Cols:   n,
+					Stride: n,
+					Data:   make([]float64, mo*n),
+				}
+				for i := 0; i < mo; i++ {
+					for j := i; j < n; j++ {
+						r.Data[i*n+j] = a.Data[i*a.Stride+j]
```

Comment:
`r.Data[i*n+j] = a.Data[(test.offset+i)*a.Stride+j]`
Then loosen the tolerance below to 1e-12 and voilà.